### PR TITLE
Remove `repos.list` from site configuration

### PR DESCRIPTION
**Pull request by tsenart**
#### Background

In transitioning to configuring repositories to be synced using only scoped external services (part of #914), the top-level site configuration option `repos.list` needs to be removed.

#### Compatibility

This change-set is *not* backwards compatible and it doesn't automatically migrate top-level repos configured with `repos.list` to be under any already configured external service.

Admins will have to manually re-configure repos that were individually added to `repos.list` to the appropriate external service configuration via the correspondent UI section. *How-to* do this will be documented and this breaking change reflected in the CHANGELOG.

We could auto-migrate `repos.list` to their respective external services, but this wouldn't succeed for all supported code hosts, since not all of them have already implemented functionality to select individual repos (e.g. `github.repos`).

Because of this, I am uncertain if it's worth the effort to do auto-migration for those code hosts that have these feature but leave out others. @nicksnyder and @sqs: Please chime in on this question.

**EDIT**: As per https://github.com/sourcegraph/sourcegraph/pull/1391#issuecomment-447040370, documentation and CHANGELOG is indeed the way to go.

#### Deployment

- **sourcegraph.com**: No special considerations.
- **Dog food**: No special considerations.
- **Customers**: Re-configuration of `repos.list` needed using external services.

#### Notes

I'm adding some question comments to the code, one per commit, that I'll seek answers for before finishing this PR. Those commits with titles beginning with `delete!` will be left out of the final change set.

### CODE REVIEW

File: cmd/repo-updater/repos/reposlist.go [hunk 1]
```diff
@@ -722,6 +722,10 @@ func (r *repoList) updateSource(source string, newList sourceRepoList) (enqueued
 var Scheduler = newUpdateScheduler()
 
 // RunRepositorySyncWorker runs the worker that syncs repositories from external code hosts to Sourcegraph
+//
```
Comment (tsenart) [inline cmd/repo-updater/repos/reposlist.go]: Any hints @nicksnyder?

File: cmd/repo-updater/repos/reposlist.go [hunk 2]
```diff
@@ -726,6 +726,10 @@ var Scheduler = newUpdateScheduler()
 // QUESTION(tsenart):
 //   Do we still rely on the old scheduler? If not, can all code related to
 //   the old scheduler be deleted?
+//
+// QUESTION(tsenart):
```
Comment (tsenart) [inline cmd/repo-updater/repos/reposlist.go]: Any hints @nicksnyder?

File: cmd/repo-updater/repos/reposlist.go (hunk 1 above)
Comment (nicksnyder) [reply cmd/repo-updater/repos/reposlist.go]: UpdateScheduler2Enabled defaults to true, so we have been using the new scheduler on dogfood for almost a month (the scheduler doesn't run on sourcegraph.com). It hasn't been tested by any customers though. I have a weak preference for keeping the old scheduler around for 1 release as a backup plan, but if the existence of that old code is inconvenient then we can move quickly and just delete that codepath. My hope is that since you are just removing code for `repos.list` that it isn't actually that inconvenient.

The new scheduler is better designed, better understood, and better tested, so I don't see a world where we would want to go back to the old scheduler, even if there were some issues with the new one (we would just fix the new one).

File: cmd/repo-updater/repos/reposlist.go (hunk 2 above)
Comment (nicksnyder) [reply cmd/repo-updater/repos/reposlist.go]: Here is the example for GitHub:

https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repos/github.go#L34:15

It calls `conf.GitHubConfigs` which either reads from the config (old code path) or queries external services (new code path). It restarts the sync worker after the config updates.

File: cmd/repo-updater/repos/reposlist.go (hunk 1 above)
Comment (tsenart) [reply cmd/repo-updater/repos/reposlist.go]: Thanks for the context. I agree with leaving it around for 1 release.

**Comment by nicksnyder** [comment]
>  I am uncertain if it's worth the effort to do auto-migration for those code hosts that have these feature but leave out others.

Almost by definition of the reason we are removing repos.list (it makes it hard to associate with individual code hosts), I don't think it makes sense to do an automatic migration since I am not sure how we could do said migration unambiguously. I think documentation/changelog is the right solution here.

**Comment by tsenart** [comment]
>  I think documentation/changelog is the right solution here.

+1

File: cmd/repo-updater/repos/reposlist.go (hunk 2 above)
Comment (tsenart) [reply cmd/repo-updater/repos/reposlist.go]: Ah, I understand now. Thanks! I hope to find that this means we can pretty much remove the whole `reposlist.go` file.

Comment (tsenart) [reply cmd/repo-updater/repos/reposlist.go]: PS: I thought there was a single worker for all different connections across code hosts. Didn't remember the changes you had made in previous PRs.

File: legacyschema/schema.go [hunk 3]
```diff
@@ -374,15 +374,16 @@ type SiteConfiguration struct {
...
-	ReposList                         []*Repository                `json:"repos.list,omitempty"`
-	ReviewBoard                       []*ReviewBoard               `json:"reviewBoard,omitempty"`
-	SearchIndexEnabled                *bool                        `json:"search.index.enabled,omitempty"`
-	SiteID                            string                       `json:"siteID,omitempty"`
-	TlsLetsencrypt                    string                       `json:"tls.letsencrypt,omitempty"`
-	TlsCert                           string                       `json:"tlsCert,omitempty"`
-	TlsKey                            string                       `json:"tlsKey,omitempty"`
-	UpdateChannel                     string                       `json:"update.channel,omitempty"`
-	UseJaeger                         bool                         `json:"useJaeger,omitempty"`
+	// QUESTION(tsenart): Should we remove ReposList from here too?
```
Comment (tsenart) [inline legacyschema/schema.go]: Any hints @nicksnyder / @slimsag?

File: CHANGELOG.md [hunk 4]
```diff
@@ -47,7 +47,23 @@ All notable changes to Sourcegraph are documented in this file.
...
+The top-level `repos.list` site configuration was removed in favour of each code-host's equivalent options,
+now configured via the new _External Services UI_ available at `/site-admin/external-services`.
+
+Equivalent options in code hosts configuration **supported today**:
+
+- Github via [`github.repos`](https://docs.sourcegraph.com/admin/site_config/all#repos-array)
+- Gitlab via [`gitlab.projectQuery`](https://docs.sourcegraph.com/admin/site_config/all#projectquery-array)
+- Phabricator via [`phabricator.repos`](https://docs.sourcegraph.com/admin/site_config/all#phabricator-array)
+
+Supported [🔜](https://github.com/sourcegraph/sourcegraph/issues/1324):
```
Comment (nicksnyder) [inline CHANGELOG.md]: I would de-emphasize promises about the future in the changelog.

```
Equivalent options in code hosts configuration:
- Github via [`github.repos`](https://docs.sourcegraph.com/admin/site_config/all#repos-array)
- Gitlab via [`gitlab.projectQuery`](https://docs.sourcegraph.com/admin/site_config/all#projectquery-array)
- Phabricator via [`phabricator.repos`](https://docs.sourcegraph.com/admin/site_config/all#phabricator-array)
- [Other code hosts](https://github.com/sourcegraph/sourcegraph/issues/1324)
```

I also think it would be ok to condense this further:
```
### Removed
 
- The `siteID` site configuration option was removed because it is no longer needed. If you previously specified this in site configuration, a new, random site ID will be generated upon server startup. You can safely remove the existing `siteID` value from your site configuration after upgrading.
- The **Info** panel was removed. The information it presented can be viewed in the hover.
- The top-level `repos.list` site configuration was removed in favour of each code-host's equivalent options,
now configured via the new _External Services UI_ available at `/site-admin/external-services`. Equivalent options in code hosts configuration:
  - Github via [`github.repos`](https://docs.sourcegraph.com/admin/site_config/all#repos-array)
  - Gitlab via [`gitlab.projectQuery`](https://docs.sourcegraph.com/admin/site_config/all#projectquery-array)
  - Phabricator via [`phabricator.repos`](https://docs.sourcegraph.com/admin/site_config/all#phabricator-array)
  - [Other code hosts](https://github.com/sourcegraph/sourcegraph/issues/1324)
```

File: legacyschema/schema.go (hunk 3 above)
Comment (nicksnyder) [reply legacyschema/schema.go]: The only reason not to move it would be because we need to do a migration, but since we aren't doing that I think it is safe to remove (and good to, to ensure all references are cleaned up).

File: CHANGELOG.md (hunk 4 above)
Comment (tsenart) [reply CHANGELOG.md]: 👍

**Comment by tsenart** [comment]
This PR is ready for review. It's a major code burn!

maxresdefault

File: cmd/repo-updater/repos/reposlist.go [hunk 5]
```diff
@@ -721,99 +715,6 @@ func (r *repoList) updateSource(source string, newList sourceRepoList) (enqueued
 // Scheduler schedules repo updates.
 var Scheduler = newUpdateScheduler()
 
-// RunRepositorySyncWorker runs the worker that syncs repositories from external code hosts to Sourcegraph
-func RunRepositorySyncWorker(ctx context.Context) {
```
Comment (nicksnyder) [inline cmd/repo-updater/repos/reposlist.go]: This shouldn't get deleted. What made you think it should?

Comment (tsenart) [reply cmd/repo-updater/repos/reposlist.go]: From what I can see, this worker only syncs ReposList defined repos.

Comment (tsenart) [reply cmd/repo-updater/repos/reposlist.go]: Ha, I see now how I misread this code. Will fix.

Comment (tsenart) [reply cmd/repo-updater/repos/reposlist.go]: No tests failed :-(

Comment (nicksnyder) [reply cmd/repo-updater/repos/reposlist.go]: The relevant lines to delete are

```
_, newMap := updateConfig(ctx, c.ReposList)
Scheduler.updateSource("internalConfig", newMap)
```

```
newList, _ := updateConfig(ctx, c.ReposList)
repos.updateSource("internalConfig", newList)
```

Everything else needs to stay. 

The code paths that sync all the other repo sources are:
```
Scheduler.run(ctx2)
```

and 

```
go repos.updateLoop(ctx2)
```

Comment (tsenart) [reply cmd/repo-updater/repos/reposlist.go]: Addressed.

Comment (nicksnyder) [reply cmd/repo-updater/repos/reposlist.go]: > No tests failed :-(

If the intent were to actually delete this function, there would be a lot more dead code to delete (e.g. scheduler.go) and tests would definitely fail (but I am glad you didn't go through that effort since we don't actually want to delete).

Our code is generally not in the state of "if you can delete code without tests failing then the code wasn't necessary to begin with"

File: doc/admin/repo/auth.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Repositories that need HTTP(S) or SSH authentication
```
Comment (nicksnyder) [inline doc/admin/repo/auth.md]: is this related to this change?

Comment (nicksnyder) [reply doc/admin/repo/auth.md]: Ah, I see it was moved/edited

Comment (tsenart) [reply doc/admin/repo/auth.md]: Yes. See this commit in isolation: https://github.com/sourcegraph/sourcegraph/pull/1391/commits/ac743d74f7b76794caff36e42a87e61c09916bf6

File: legacyschema/site.schema.json [hunk 7]
```diff
@@ -232,13 +232,6 @@
       "type": "integer",
       "default": 5
     },
-    "repos.list": {
-      "description": "JSON array of configuration for external repositories.",
-      "type": "array",
-      "items": {
-        "$ref": "#/definitions/Repository"
```
Comment (nicksnyder) [inline legacyschema/site.schema.json]: Can this definition be deleted too?

Comment (tsenart) [reply legacyschema/site.schema.json]: I don't know but I'll take a look to see if we can!

Comment (tsenart) [reply legacyschema/site.schema.json]: I removed it and tests pass. What do you think?

Comment (nicksnyder) [reply legacyschema/site.schema.json]: It would fail to build if it were a problem, so this should be fine.

**Comment by tsenart** [comment]
Alright, waiting for @slimsag to review the `legacyschema` changes.

**Comment by slimsag** [comment]
This change made me realize we have lots of breaking changes we will need to communicate clearly to our users. The CHANGELOG is good, but hard to miss things. So I am proposing to create a guide specifically for upgrading from 2.x to 3.0: https://github.com/sourcegraph/sourcegraph/issues/1426

File: doc/integration/bitbucket_server.md [hunk 8]
```diff
@@ -2,8 +2,6 @@
 
 Sourcegraph integrates with Bitbucket Server and Bitbucket Data Center.
 
-> NOTE: Support for [Bitbucket Cloud](https://bitbucket.org) is coming soon. In the meantime, you can add Bitbucket Cloud repositories using [`repos.list`](../admin/repo/add_from_git_repository.md).
```
Comment (slimsag) [inline doc/integration/bitbucket_server.md]: We should probably retain the 'coming soon' part of this message?

Comment (slimsag) [reply doc/integration/bitbucket_server.md]: Or maybe not? I'm not sure. @sqs?

Comment (tsenart) [reply doc/integration/bitbucket_server.md]: Is it actually coming soon?

Comment (slimsag) [reply doc/integration/bitbucket_server.md]: Yeah, disregard, this should be in an issue anyway.
